internal/controller/notificationservice_controller_test.go

```diff
@@ -90,120 +95,199 @@ var _ = Describe("NotificationService Controller", func() {
-						Results: []tektonv1.PipelineRunResult{
-							{
-								Name:  "IMAGE_DIGEST",
-								Value: *tektonv1.NewStructuredValues("image_digest_value"),
-							},
-							{
-								Name:  "IMAGE_URL",
-								Value: *tektonv1.NewStructuredValues("image"),
-							},
-							{
-								Name:  "CHAINS-GIT_URL",
-								Value: *tektonv1.NewStructuredValues("git_url_value"),
-							},
-							{
-								Name:  "CHAINS-GIT_COMMIT",
-								Value: *tektonv1.NewStructuredValues("git_commit_value"),
+					By("Updating status to completed successfully")
+					createdPipelineRun.Status = tektonv1.PipelineRunStatus{
+						PipelineRunStatusFields: tektonv1.PipelineRunStatusFields{
+							StartTime:      &metav1.Time{Time: time.Now()},
+							CompletionTime: &metav1.Time{Time: time.Now().Add(5 * time.Minute)},
+							Results: []tektonv1.PipelineRunResult{
+								{
+									Name:  "IMAGE_DIGEST",
+									Value: *tektonv1.NewStructuredValues("image_digest_value"),
+								},
+								{
+									Name:  "IMAGE_URL",
+									Value: *tektonv1.NewStructuredValues("image"),
+								},
+								{
+									Name:  "CHAINS-GIT_URL",
+									Value: *tektonv1.NewStructuredValues("git_url_value"),
+								},
+								{
+									Name:  "CHAINS-GIT_COMMIT",
+									Value: *tektonv1.NewStructuredValues("git_commit_value"),
+								},
 							},
 						},
-					},
-					Status: v1.Status{
-						Conditions: v1.Conditions{
-							apis.Condition{
-								Message: "Tasks Completed: 12 (Failed: 0, Cancelled 0), Skipped: 2",
-								Reason:  "Completed",
-								Status:  "True",
-								Type:    apis.ConditionSucceeded,
+						Status: v1.Status{
+							Conditions: v1.Conditions{
+								apis.Condition{
+									Message: "Tasks Completed: 12 (Failed: 0, Cancelled 0), Skipped: 2",
+									Reason:  "Completed",
+									Status:  "True",
+									Type:    apis.ConditionSucceeded,
+								},
 							},
 						},
-					},
-				}
-				Expect(k8sClient.Status().Update(ctx, createdPipelineRun)).Should(Succeed())
+					}
+					Expect(k8sClient.Status().Update(ctx, createdPipelineRun)).Should(Succeed())
+					// The pipelinerun should be reconciled:
+					// Read the results, add the notification annotation, remove the finalizer
+					Eventually(func() bool {
+						err := k8sClient.Get(ctx, pushPipelineRunLookupKey, createdPipelineRun)
+						Expect(err).ToNot(HaveOccurred())
+						return metadata.HasAnnotationWithValue(createdPipelineRun, NotificationPipelineRunAnnotation, NotificationPipelineRunAnnotationValue)
+					}, timeout, interval).Should(BeTrue())
+					Expect(controllerutil.ContainsFinalizer(createdPipelineRun, NotificationPipelineRunFinalizer)).To(BeFalse())
+					// Check the Notify was called only once
+					Expect(mn.Counter).To(Equal(1))
+					// Check that notifications metric increased by 1
+					Expect(testutil.ToFloat64(notifications)).To(Equal(notificationsCounter + 1))
```

Comment:
Is it possible here that the reconciliation will run more than once between the time we get the before value and the time we get the after value?

Reply:
Because the tests are separated and no other pipelines run, there should not be other reconciliations than the one that we currently testing.

I thought about adding **pipelinerun name** and **namespace** labels to the counters, and in the test we can check that a metric with these specific labels is 1.
I need to think if it will add us value in the metrics and alerts

Reply:
Let's not create additional data series (i.e. let's not add labels). It could have a big effect on storage.